### PR TITLE
Allow null types

### PR DESCRIPTION
**Pull request by veronicagg**
Addressing https://github.com/Azure/oav/issues/155
@amarzavery please take  a look when you get a chance, I'll be adding tests, but wanted to send a PR with the modified code, since we talked about this. 
I changed the condition on when nullable applies a bit, so want to know if you agree or we should be doing it differently. 
Basically, I noticed that we have some specs that specifically set x-nullable to false, in which case I think that even if the property is not required, we should probably honor that setting. 
Logic is now that if x-nullable is not undefined then we honor the value, and if the property is undefined, then we look at whether the property is required or not. Thoughts?

### CODE REVIEW

**Comment by amarzavery** [comment]
@veronicagg - Please resolve conflicts with upstream master

File: lib/util/utils.js [hunk 1]
```diff
@@ -580,6 +580,39 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
   return model;
 }
 
+exports.allowNullType = function allowNullType(entity, isPropRequired) {
```
[the remark this answers is not in the document]
Comment (amarzavery) [reply lib/util/utils.js]: The valueType could be a primitive type or an inline object with properties or a reference to another object (composite type) or an array type.
- If it is a reference to another type then it is fine.
- If it is an inline object with properties then we need to process it // allowNullableTypes()
- If it is a primitive type like string, number, boolean then we need to process it // allowNullTypes()
- If it is an array then we need to process it

**Comment by veronicagg** [comment]
@amarzavery I've updated the code to account for the cases we've talked about and added tests, let me know what you think.
I see the CI is failing for other tests, I need to look into that, I wonder if my merge didn't work well, as I see something failing for x-ms-parameterized-host.

**Comment by amarzavery** [comment]
@veronicagg - Please set
`options.shouldResolveNullableTypes = false` for semantic validation over [here](https://github.com/Azure/oav/blob/master/lib/validate.js#L68)

File: lib/commands/resolve-spec.js [hunk 2]
```diff
@@ -53,6 +53,12 @@ exports.builder = {
     boolean: true,
     default: false
   },
+  n: {
+    alias: 'nullable',
+    describe: 'Should nullable be resolved?',
```
Comment (amarzavery) [inline lib/commands/resolve-spec.js]: Should nullable **types** be resolved?

File: lib/util/utils.js [hunk 3]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
+    if (typeof entity['x-nullable'] === 'string') {
+      if (entity['x-nullable'].toLowerCase() === 'false') {
+        entity['x-nullable'] = false;
+      } else if (entity['x-nullable'].toLowerCase() === 'true') {
+        entity['x-nullable'] = true;
+      }
+    }
+
+    let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
+    if ((isPropNullable === undefined && !isPropRequired) || isPropNullable) {
```
Comment (amarzavery) [inline lib/util/utils.js]: It is worthwhile providing a comment or may be putting a True False logic table that we had drawn on the white board in the comments over here. That will be helpful to us in the future.

File: lib/util/utils.js [hunk 4]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
+      let savedEntity = entity
```
Comment (amarzavery) [inline lib/util/utils.js]: // Handling nullable parameters

File: lib/util/utils.js [hunk 5]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
+      if (savedEntity.in) {
+        entity.oneOf = [{ "type": entity.type }, { "type": "null" }];
+        delete entity.type
+
+      }
+      else {
```
Comment (amarzavery) [inline lib/util/utils.js]: as per JS code style, else should be on the same line as the if closing bracket
```
} else {
 . ..
}
```

File: lib/util/utils.js [hunk 6]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
+      }
+    }
+
+    let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
+    if ((isPropNullable === undefined && !isPropRequired) || isPropNullable) {
+      let savedEntity = entity
+      if (savedEntity.in) {
+        entity.oneOf = [{ "type": entity.type }, { "type": "null" }];
+        delete entity.type
+
```
Comment (amarzavery) [inline lib/util/utils.js]: remove extra line

File: lib/util/utils.js [hunk 7]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
+      else {
+        entity = {};
+        entity.oneOf = [savedEntity, { "type": "null" }];
+      }
+    }
+  }
+
+  // if there's a $ref
+  if (entity && entity["$ref"]) {
     let isPropNullable = entity['x-nullable'] && typeof entity['x-nullable'] === 'boolean';
```
Comment (amarzavery) [inline lib/util/utils.js]: duplicate code, can we extract this into a function and call that function?

File: lib/util/utils.js [hunk 8]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
     let modelProperties = model.properties;
     for (let propName in modelProperties) {
-      let isPropRequired = model.required ? model.required.some((p) => { return p == propName; }) : false
-      if (modelProperties[propName].properties) {
+      let isPropRequired = model.required ? model.required.some((p) => { return p == propName; }) : false;
+      // process properties if present
+      if (modelProperties[propName].properties || modelProperties[propName].additionalProperties) {
         modelProperties[propName] = exports.allowNullableTypes(modelProperties[propName]);
       }
       else {
```
Comment (amarzavery) [inline lib/util/utils.js]: else on the same line

File: lib/util/utils.js [hunk 9]
```diff
@@ -580,39 +580,133 @@ exports.relaxModelLikeEntities = function relaxModelLikeEntities(model) {
...
     }
   }
-  //this needs to happen at the end, as model is modified in allowNullType to have only oneOf value with the original contents or null
-  model = exports.allowNullType(model);
+
+  if (model && model.type) {
+    if (model.type == 'array') {
+      if (model.items) {
+        // if items object contains inline properties
+        if (model.items.properties) {
```
Comment (amarzavery) [inline lib/util/utils.js]: what about model.items.additionalProperties?

File: lib/validators/specResolver.js [hunk 10]
```diff
@@ -621,13 +621,26 @@ class SpecResolver {
     //scan definitions and properties of every model in definitions
     for (let defName in definitions) {
       let model = definitions[defName];
-      utils.allowNullableTypes(model);
+      definitions[defName] = utils.allowNullableTypes(model);
```
Comment (amarzavery) [inline lib/validators/specResolver.js]: wouldn't this make the definition itself a `oneOf` of that type and null?

**Comment by veronicagg** [comment]
Still pending some code review feedback, and full validation on azure-rest-api-specs repo.

**Comment by amarzavery** [comment]
everything LGTM. Bump the version in package.json and also update the changelog.md

**Comment by amarzavery** [comment]
This [StackOverflow link](https://stackoverflow.com/questions/47374980/schema-object-without-a-type-attribute-in-swagger-2-0) provides good explanation about requiredness `type` in a json document.

**Pull request merged**